File: src/main/java/pl/yalgrin/playnite/simplesync/mapper/objects/FilterPresetMapper.java
package pl.yalgrin.playnite.simplesync.mapper.objects;

import io.r2dbc.postgresql.codec.Json;
import io.vavr.control.Try;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import org.apache.commons.lang3.Strings;
import org.springframework.stereotype.Component;
import pl.yalgrin.playnite.simplesync.dto.filter.FilterPresetSettingsDTO;
import pl.yalgrin.playnite.simplesync.dto.filter.IdItemPropertiesDTO;
import pl.yalgrin.playnite.simplesync.dto.filter.IntItemPropertiesDTO;
import pl.yalgrin.playnite.simplesync.dto.filter.StringItemPropertiesDTO;
import pl.yalgrin.playnite.simplesync.dto.objects.FilterPresetDTO;
import pl.yalgrin.playnite.simplesync.library.domain.FilterPreset;
import tools.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

@Component
@RequiredArgsConstructor
public class FilterPresetMapper extends AbstractObjectMapper<FilterPreset, FilterPresetDTO> {

    private final ObjectMapper objectMapper;

    @SneakyThrows
    @Override
    public FilterPreset fillEntity(FilterPresetDTO dto, FilterPreset target) {
        dto.setRemoved(false);
        byte[] newDataToSave = objectMapper.writeValueAsBytes(dto);

        target.setChanged(
                !Strings.CS.equals(target.getName(), dto.getName()) || target.isRemoved() || isPreviousDataDifferent(
                        target, dto));
        target.setName(dto.getName());
        target.setSavedData(Json.of(newDataToSave));
        target.setRemoved(false);
        return target;
    }

    private boolean isPreviousDataDifferent(FilterPreset target, FilterPresetDTO dto) {
        byte[] previousSavedData = Optional.ofNullable(target.getSavedData()).map(Json::asArray).orElse(null);
        FilterPresetDTO previousDTO = Try.of(() -> objectMapper.readValue(previousSavedData, FilterPresetDTO.class))
                .getOrNull();
        return previousDTO == null
                || !Strings.CS.equals(previousDTO.getId(), dto.getId())
                || !Strings.CS.equals(previousDTO.getName(), dto.getName())
                || !Objects.equals(previousDTO.isRemoved(), dto.isRemoved())
                || !Strings.CS.equals(previousDTO.getSortingOrder(), dto.getSortingOrder())
                || !Strings.CS.equals(previousDTO.getSortingOrderDirection(), dto.getSortingOrderDirection())
                || !Strings.CS.equals(previousDTO.getGroupingOrder(), dto.getGroupingOrder())
                || !Objects.equals(previousDTO.isShowInFullscreenQuickSelection(),
                dto.isShowInFullscreenQuickSelection())
                || areSettingsDifferent(previousDTO.getSettings(), dto.getSettings());
    }

    private boolean areSettingsDifferent(FilterPresetSettingsDTO previousDTO, FilterPresetSettingsDTO newDTO) {
        return (previousDTO == null) != (newDTO == null)
                || (previousDTO != null && (
                !Objects.equals(previousDTO.isUseAndFilteringStyle(), newDTO.isUseAndFilteringStyle())
                        || !Objects.equals(previousDTO.isInstalled(), newDTO.isInstalled())
                        || !Objects.equals(previousDTO.isUninstalled(), newDTO.isUninstalled())
                        || !Objects.equals(previousDTO.isHidden(), newDTO.isHidden())
                        || !Objects.equals(previousDTO.isFavorite(), newDTO.isFavorite())
                        || !Strings.CS.equals(previousDTO.getName(), newDTO.getName())
                        || !Strings.CS.equals(previousDTO.getVersion(), newDTO.getVersion())
                        || areItemPropertiesDifferent(previousDTO.getReleaseYear(), newDTO.getReleaseYear())
                        || areItemPropertiesDifferent(previousDTO.getGenre(), newDTO.getGenre())
                        || areItemPropertiesDifferent(previousDTO.getPlatform(), newDTO.getPlatform())
                        || areItemPropertiesDifferent(previousDTO.getDeveloper(), newDTO.getDeveloper())
                        || areItemPropertiesDifferent(previousDTO.getCategory(), newDTO.getCategory())
                        || areItemPropertiesDifferent(previousDTO.getTag(), newDTO.getTag())
                        || areItemPropertiesDifferent(previousDTO.getSeries(), newDTO.getSeries())
                        || areItemPropertiesDifferent(previousDTO.getRegion(), newDTO.getRegion())
                        || areItemPropertiesDifferent(previousDTO.getSource(), newDTO.getSource())
                        || areItemPropertiesDifferent(previousDTO.getAgeRating(), newDTO.getAgeRating())
                        || areItemPropertiesDifferent(previousDTO.getLibrary(), newDTO.getLibrary())
                        || areItemPropertiesDifferent(previousDTO.getCompletionStatuses(),
                        newDTO.getCompletionStatuses())
                        || areItemPropertiesDifferent(previousDTO.getFeature(), newDTO.getFeature())
                        || areItemPropertiesDifferent(previousDTO.getUserScore(), newDTO.getUserScore())
                        || areItemPropertiesDifferent(previousDTO.getCriticScore(), newDTO.getCriticScore())
                        || areItemPropertiesDifferent(previousDTO.getCommunityScore(), newDTO.getCommunityScore())
                        || areItemPropertiesDifferent(previousDTO.getLastActivity(), newDTO.getLastActivity())
                        || areItemPropertiesDifferent(previousDTO.getRecentActivity(), newDTO.getRecentActivity())
                        || areItemPropertiesDifferent(previousDTO.getAdded(), newDTO.getAdded())
                        || areItemPropertiesDifferent(previousDTO.getModified(), newDTO.getModified())
                        || areItemPropertiesDifferent(previousDTO.getPlayTime(), newDTO.getPlayTime())
                        || areItemPropertiesDifferent(previousDTO.getInstallSize(), newDTO.getInstallSize())
        ));
    }

    private boolean areItemPropertiesDifferent(StringItemPropertiesDTO previousDTO, StringItemPropertiesDTO newDTO) {
        return (previousDTO == null) != (newDTO == null)
                || (previousDTO != null && areListsDifferent(previousDTO.getValues(), newDTO.getValues()));
    }

    private boolean areItemPropertiesDifferent(IdItemPropertiesDTO previousDTO, IdItemPropertiesDTO newDTO) {
        return (previousDTO == null) != (newDTO == null)
                || (previousDTO != null && (areListsDifferent(previousDTO.getIds(), newDTO.getIds())
                || !Strings.CS.equals(previousDTO.getText(), newDTO.getText())));
    }

    private boolean areItemPropertiesDifferent(IntItemPropertiesDTO previousDTO, IntItemPropertiesDTO newDTO) {
        return (previousDTO == null) != (newDTO == null)
                || (previousDTO != null && areListsDifferent(previousDTO.getValues(), newDTO.getValues()));
    }

    private <T> boolean areListsDifferent(List<T> previousList, List<T> newList) {
        Integer previousCount = Optional.ofNullable(previousList).map(List::size).orElse(0);
        Integer newCount = Optional.ofNullable(newList).map(List::size).orElse(0);
        if (!Objects.equals(previousCount, newCount)) {
            return true;
        }
        if (previousList != null && newList != null) {
            List<T> newListCopy = new ArrayList<>(newList);
            for (T t : previousList) {
                if (!newListCopy.remove(t)) {
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    protected FilterPresetDTO createDTO() {
        return new FilterPresetDTO();
    }

    @SneakyThrows
    @Override
    public FilterPresetDTO toDTO(FilterPreset entity) {
        FilterPresetDTO dto = objectMapper.readValue(entity.getSavedData()
                .asArray(), FilterPresetDTO.class);
        dto.setId(entity.getPlayniteId());
        dto.setName(entity.getName());
        dto.setRemoved(entity.isRemoved());
        fillDtoFields(dto, entity);
        return dto;
    }
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/library/mapper/FilterPresetMapper.kt
package pl.yalgrin.playnite.simplesync.library.mapper

import org.apache.commons.lang3.Strings
import org.springframework.stereotype.Component
import pl.yalgrin.playnite.simplesync.common.util.asJson
import pl.yalgrin.playnite.simplesync.library.domain.FilterPreset
import pl.yalgrin.playnite.simplesync.library.domain.extractDbModelOrEmpty
import pl.yalgrin.playnite.simplesync.library.dto.FilterPresetDTO
import pl.yalgrin.playnite.simplesync.library.dto.FilterPresetDatabaseModel
import pl.yalgrin.playnite.simplesync.library.dto.filter.FilterPresetSettingsDTO
import pl.yalgrin.playnite.simplesync.library.dto.filter.IdItemPropertiesDTO
import pl.yalgrin.playnite.simplesync.library.dto.filter.IntItemPropertiesDTO
import pl.yalgrin.playnite.simplesync.library.dto.filter.StringItemPropertiesDTO
import reactor.core.publisher.Mono

@Component
class FilterPresetMapper : LibraryObjectMapperImpl<FilterPreset, FilterPresetDTO>() {

    override fun fillOtherFields(entity: FilterPreset, dto: FilterPresetDTO): Mono<FilterPreset> {
        return entity.extractDbModelOrEmpty()
            .map { fillDbModelFields(it, dto) }
            .flatMap { dbModel -> dbModel.asJson() }
            .map { json -> entity.savedData = json }
            .thenReturn(entity)
    }

    private fun fillDbModelFields(dbModel: FilterPresetDatabaseModel, dto: FilterPresetDTO): FilterPresetDatabaseModel {
        dbModel.sortingOrder = dto.sortingOrder
        dbModel.sortingOrderDirection = dto.sortingOrderDirection
        dbModel.groupingOrder = dto.groupingOrder
        dbModel.showInFullscreenQuickSelection = dto.showInFullscreenQuickSelection
        dbModel.settings = dto.settings
        return dbModel
    }

    override fun hasChangedMono(dto: FilterPresetDTO, target: FilterPreset): Mono<Boolean> {
        return super.hasChangedMono(dto, target)
            .flatMap { changed -> if (changed) Mono.just(true) else isPreviousDataDifferent(target, dto) }
    }

    private fun isPreviousDataDifferent(target: FilterPreset, dto: FilterPresetDTO): Mono<Boolean> {
        return target.extractDbModelOrEmpty()
            .map { previousDTO ->
                !Strings.CS.equals(
                    previousDTO.sortingOrder,
                    dto.sortingOrder
                ) || !Strings.CS.equals(
                    previousDTO.sortingOrderDirection,
                    dto.sortingOrderDirection
                ) || !Strings.CS.equals(
                    previousDTO.groupingOrder,
                    dto.groupingOrder
                ) || (previousDTO.showInFullscreenQuickSelection != dto.showInFullscreenQuickSelection) || areSettingsDifferent(
                    previousDTO.settings,
                    dto.settings
                )
            }
            .defaultIfEmpty(true)
    }

    private fun areSettingsDifferent(previousDTO: FilterPresetSettingsDTO?, newDTO: FilterPresetSettingsDTO?): Boolean {
        return (previousDTO == null) != (newDTO == null)
                || (previousDTO != null && ((previousDTO.useAndFilteringStyle != newDTO!!.useAndFilteringStyle) || (previousDTO.isInstalled != newDTO.isInstalled) || (previousDTO.isUninstalled != newDTO.isUninstalled) || (previousDTO.isHidden != newDTO.isHidden) || (previousDTO.isFavorite != newDTO.isFavorite) || !Strings.CS.equals(
            previousDTO.name,
            newDTO.name
        ) || !Strings.CS.equals(
            previousDTO.version,
            newDTO.version
        ) || areItemPropertiesDifferent(previousDTO.releaseYear, newDTO.releaseYear)
                || areItemPropertiesDifferent(previousDTO.genre, newDTO.genre)
                || areItemPropertiesDifferent(previousDTO.platform, newDTO.platform)
                || areItemPropertiesDifferent(previousDTO.developer, newDTO.developer)
                || areItemPropertiesDifferent(previousDTO.category, newDTO.category)
                || areItemPropertiesDifferent(previousDTO.tag, newDTO.tag)
                || areItemPropertiesDifferent(previousDTO.series, newDTO.series)
                || areItemPropertiesDifferent(previousDTO.region, newDTO.region)
                || areItemPropertiesDifferent(previousDTO.source, newDTO.source)
                || areItemPropertiesDifferent(previousDTO.ageRating, newDTO.ageRating)
                || areItemPropertiesDifferent(previousDTO.library, newDTO.library)
                || areItemPropertiesDifferent(
            previousDTO.completionStatuses,
            newDTO.completionStatuses
        )
                || areItemPropertiesDifferent(previousDTO.feature, newDTO.feature)
                || areItemPropertiesDifferent(previousDTO.userScore, newDTO.userScore)
                || areItemPropertiesDifferent(previousDTO.criticScore, newDTO.criticScore)
                || areItemPropertiesDifferent(previousDTO.communityScore, newDTO.communityScore)
                || areItemPropertiesDifferent(previousDTO.lastActivity, newDTO.lastActivity)
                || areItemPropertiesDifferent(previousDTO.recentActivity, newDTO.recentActivity)
                || areItemPropertiesDifferent(previousDTO.added, newDTO.added)
                || areItemPropertiesDifferent(previousDTO.modified, newDTO.modified)
                || areItemPropertiesDifferent(previousDTO.playTime, newDTO.playTime)
                || areItemPropertiesDifferent(previousDTO.installSize, newDTO.installSize)
                ))
    }

    private fun areItemPropertiesDifferent(
        previousDTO: StringItemPropertiesDTO?,
        newDTO: StringItemPropertiesDTO?
    ): Boolean {
        return (previousDTO == null) != (newDTO == null)
                || (previousDTO != null && areListsDifferent(previousDTO.values, newDTO!!.values))
    }

    private fun areItemPropertiesDifferent(previousDTO: IdItemPropertiesDTO?, newDTO: IdItemPropertiesDTO?): Boolean {
        return (previousDTO == null) != (newDTO == null)
                || (previousDTO != null && (areListsDifferent(previousDTO.ids, newDTO!!.ids)
                || !Strings.CS.equals(previousDTO.text, newDTO.text)))
    }

    private fun areItemPropertiesDifferent(previousDTO: IntItemPropertiesDTO?, newDTO: IntItemPropertiesDTO?): Boolean {
        return (previousDTO == null) != (newDTO == null)
                || (previousDTO != null && areListsDifferent(previousDTO.values, newDTO!!.values))
    }

    private fun <T> areListsDifferent(previousList: List<T>, newList: List<T>): Boolean {
        val previousCount = previousList.size
        val newCount = newList.size
        if (previousCount != newCount) {
            return true
        }
        val newListCopy: MutableList<T> = ArrayList(newList)
        for (t in previousList) {
            if (!newListCopy.remove(t)) {
                return true
            }
        }
        return false
    }

    override fun createDTO(): FilterPresetDTO = FilterPresetDTO()

    override fun fillOtherDtoFields(dto: FilterPresetDTO, entity: FilterPreset): Mono<FilterPresetDTO> {
        return entity.extractDbModelOrEmpty()
            .map { dbModel ->
                dto.sortingOrder = dbModel.sortingOrder
                dto.sortingOrderDirection = dbModel.sortingOrderDirection
                dto.groupingOrder = dbModel.groupingOrder
                dto.showInFullscreenQuickSelection = dbModel.showInFullscreenQuickSelection
                dto.settings = dbModel.settings
                dto
            }
            .thenReturn(dto)
    }
}